File: DailyLeetcode/10Aug/ReorderedPowerOf2.java
//Solved using Sorting
//This solution checks if the digits of the number can be rearranged to form a power of two.
//Time Complexity: O(n log n)
//Space Complexity: O(1)
class Solution {
    public boolean reorderedPowerOf2(int n) {
        String target = sortedString(n);
        for (int i = 0; i < 31; i++) {
            if (sortedString(1 << i).equals(target)) return true;
        }
        return false;
    }
    private String sortedString(int x) {
        char[] arr = String.valueOf(x).toCharArray();
        Arrays.sort(arr);
        return new String(arr);
    }
}

File: DailyLeetcode/10Aug/ReorderedPowerOf2.cpp
//Time complexity=O(logn)
//Space complexity=O(1)
class Solution {
public:
 // Count digit frequencies of a non-negative integer n
    vector<int> freq(int n){
        vector<int>res(10,0);
        while(n>0){
            res[n%10]++;
            n=n/10;
        }
        return res;
    }
    // Compare two frequency arrays (size 10) for equality
    bool compa(const vector<int>&a ,const vector<int>&b){
        for(int i=0;i<10;i++){
            if(a[i]!=b[i]){
                return false;
            }
        }
        return true;
    }

    bool reorderedPowerOf2(int n) {
        // Frequency of digits in the input number n
        vector<int>freq2=freq(n);
         // Check all powers of two up to 2^30 (fits in 32-bit signed int ie till limit ie 10^9)
        for(int i=0;i<30;i++){
            int pow2=(int)pow(2,i); //2^i
            if(compa(freq2,freq(pow2))){
                return true;
            }
        }
        return false;
        
    }
};